swift/ql/src/queries/Security/CWE-760/ConstantSalt.ql

```diff
@@ -0,0 +1,63 @@
+/**
+ * @name Constant salt
+ * @description Using constant salts for password hashing is not secure, because potential attackers can pre-compute the hash value via dictionary attacks.
+ * @kind path-problem
+ * @problem.severity error
+ * @security-severity 7.5
+ * @precision high
+ * @id swift/constant-salt
+ * @tags security
+ *       external/cwe/cwe-760
+ */
+
+import swift
+import codeql.swift.dataflow.DataFlow
+import codeql.swift.dataflow.TaintTracking
+import codeql.swift.dataflow.FlowSteps
+import DataFlow::PathGraph
+
+/**
+ * A constant salt is created through either a byte array or string literals.
+ */
+class ConstantSaltSource extends Expr {
+  ConstantSaltSource() {
+    this = any(ArrayExpr arr | arr.getType().getName() = "Array<UInt8>") or
+    this instanceof StringLiteralExpr
+  }
+}
+
+/**
+ * A class for all ways to use a constant salt.
+ */
+class ConstantSaltSink extends Expr {
+  ConstantSaltSink() {
+    // `salt` arg in `init` is a sink
+    exists(ClassOrStructDecl c, AbstractFunctionDecl f, CallExpr call, int arg |
+      c.getFullName() = ["HKDF", "PBKDF1", "PBKDF2", "Scrypt"] and
+      c.getAMember() = f and
+      f.getName().matches("%init(%salt:%") and
```

Comment:
This looks like a great start, my only concern is how often salts are specified in the wild using the functions modelled above.  I did a quick MRVA query for parameters called "salt" (or similar) and the most common call that appears to be relevant is to a thing called `CCKeyDerivationPBKDF(_:_:_:_:_:_:_:_:_:)` from `CommonCrypto`.  Do you think we should perhaps create a follow-up issue to add a model of that to this query as well?

Reply:
We should add further support for `CommonCrytpo`, not just for this query, but perhaps other crypto queries too. This might be a good first step in supporting other crypto APIs.

Reply:
Would you mind creating an issue for that, listing the support you think we need to add?

Reply:
Done. Just added one with an initial set of tasks along with some useful resources.